security/pkg/k8s/tokenreview/k8sauthn.go

```diff
@@ -16,22 +16,56 @@ package tokenreview
 
 import (
 	"context"
+	"encoding/base64"
+	"encoding/json"
 	"fmt"
 	"strings"
 
 	k8sauth "k8s.io/api/authentication/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 
-	"istio.io/istio/pkg/jwt"
+	"istio.io/pkg/env"
+	"istio.io/pkg/log"
 )
 
-const (
+var (
 	// The default audience for SDS trustworthy JWT. This is to make sure that the CSR requests
 	// contain the JWTs intended for Citadel.
-	DefaultAudience = "istio-ca"
+	DefaultAudience = env.RegisterStringVar("TOKEN_AUDIENCE", "istio-ca", "Audience to check in accepted JWTW tokens")
+	RequireAudience = env.RegisterBoolVar("REQUIRE_AUDIENCE", false, "Reject tokens without audience. If false, K8S token will be accepted")
 )
 
+type jwtPayload struct {
+	// Aud is the expected audience, defaults to istio-ca - but is based on istiod.yaml configuration.
+	// If set to a different value - use the value defined by istiod.yaml. Env variable can
+	// still override
+	Aud []string `json:"aud"`
+}
+
+// checkAudience detects if the token has an audience or is a 1st party JWT.
+// This allows migration and interop - we need to accept both.
+func checkAudience(jwt string) bool {
```

Comment:
Shall we call it audienceExist? Can you add a test for the function?

Reply:
It is covered by the current integration tests. I can add a test as well - this probably needs to be refactored a bit as well..